demonstrations/tutorial_qutrits_bernstein_vazirani.py

```diff
@@ -0,0 +1,394 @@
+   :align: center
+
+   Bernstein Vazirani algorithm.
+
+
+What we can see is that by simply using Hadamard gates before and after the oracle, after a single run the output of the circuit is exactly the hidden value of :math:`\vec{a}`. Let's do a little math to verify that this is so.
+
+- First, the input to our circuit is :math:`|0001\rangle`.
+
+- The second step is to apply Hadamard gates on this state, and for this we must use the following property
+
+  .. math::
+        H^n|\vec{x}\rangle = \frac{1}{\sqrt{2^n}}\sum_{\vec{z} \in \{0,1\}^n}(-1)^{\vec{x}\cdot\vec{z}}|\vec{z}\rangle.
+
+  This expression may seem complicated to obtain but understanding the case of one qubit, we would have to multiply and group. It is simply to realize that :math:`H|x\rangle = \frac{|0\rangle+(-1)^x |1\rangle}{\sqrt{2}}` where :math:`x \in \{0.1\}`.
+
+  Taking as input the value :math:`|0001\rangle`, we obtain the state
+
+  .. math::
+        |\phi_1\rangle:=H^4|0001\rangle = H^3|000\rangle\otimes H|1\rangle = \frac{1}{\sqrt{2^3}}\left(\sum_{z \in \{0,1\}^3}|\vec{z}\rangle\right)\left(\frac{|0\rangle-|1\rangle}{\sqrt{2}}\right).
+
+  As you can see, we have separated the first three qubits from the fourth for simplicity.
+
+- If we now apply our operator :math:`U_f`,
+
+  .. math::
+      |\phi_2\rangle:= U_f |\phi_1\rangle = \frac{1}{\sqrt{2^3}}\left(\sum_{\vec{z} \in \{0,1\}^3}|\vec{z}\rangle\frac{|\vec{a}\cdot\vec{z} \pmod 2\rangle-|1 + \vec{a}\cdot\vec{z} \pmod 2\rangle}{\sqrt{2}}\right).
+
+  Depending on the value of :math:`f(\vec{x})` the final part of the expression can take two values and it can be checked that
+
+  .. math::
+      |\phi_2\rangle = \frac{1}{\sqrt{2^3}}\left(\sum_{\vec{z} \in \{0,1\}^3}|\vec{z}\rangle(-1)^{\vec{a}\cdot\vec{z}}\frac{|0\rangle-|1\rangle}{\sqrt{2}}\right).
+
+  This is because if :math:`\vec{a}\cdot\vec{z}` takes the value :math:`0`, we will have the :math:`\frac{|0\rangle - |1\rangle}{\sqrt{2}}` and if it takes the value :math:`1`, the result will be :math:`\frac{|1\rangle - |0\rangle}{\sqrt{2}} = - \frac{|0\rangle - |1\rangle}{\sqrt{2}}`. Therefore, by calculating :math:`(-1)^{\vec{a}\cdot\vec{z}}` we cover both cases.
+
+- After this, we can enter the :math:`(-1)^{\vec{a}\cdot\vec{z}}` factor in the :math:`|\vec{z}\rangle` term and disregard the last qubit since we are not going to use it again
+
+  .. math::
+        |\phi_2\rangle =\frac{1}{\sqrt{2^3}}\sum_{\vec{z} \in \{0,1\}^3}(-1)^{\vec{a}\cdot\vec{z}}|\vec{z}\rangle.
+
+  Note that you cannot always disregard a qubit. In cases where there is entanglement with other qubits it is not possible but in this case it is separable.
+
+- Finally, we will reapply the property of the first step to calculate the result after using the Hadamard
+
+  .. math::
+        |\phi_3\rangle := H^3|\phi_2\rangle = \frac{1}{2^3}\sum_{\vec{z} \in \{0,1\}^3}(-1)^{\vec{a}\cdot\vec{z}}\left(\sum_{\vec{y} \in \{0,1\}^3}(-1)^{\vec{z}\cdot\vec{y}}|\vec{y}\rangle\right).

```

Comment:
```suggestion
        |\phi_3\rangle := H^{\otimes 3}|\phi_2\rangle = \frac{1}{2^3}\sum_{\vec{z} \in \{0,1\}^3}(-1)^{\vec{a}\cdot\vec{z}}\left(\sum_{\vec{y} \in \{0,1\}^3}(-1)^{\vec{z}\cdot\vec{y}}|\vec{y}\rangle\right).

```

Please check all instances of this to ensure the tensor product is included in the definitions (otherwise it reads as just applying the Hadamard multiple times!)